From 2fde4db851cdded367500397e83263f0373692c7 Mon Sep 17 00:00:00 2001
From: Zhang Heng <zhangheng@mprc.pku.edu.cn>
Date: Thu, 18 Jan 2018 15:56:42 +0800
Subject: [PATCH 20/21] fix typo: swap_foreground_background_csr

Signed-off-by: Zhang Heng <zhangheng@mprc.pku.edu.cn>
---
 target-riscv/op_helper.c | 23 +++++++++++------------
 1 file changed, 11 insertions(+), 12 deletions(-)

diff --git a/target-riscv/op_helper.c b/target-riscv/op_helper.c
index fcb1f50..405b863 100644
--- a/target-riscv/op_helper.c
+++ b/target-riscv/op_helper.c
@@ -734,7 +734,7 @@ void set_privilege(CPURISCVState *env, target_ulong newpriv)
     env->priv = newpriv;
 }
 
-#define swap_background_background_csr(env, csr_no)                     \
+#define swap_foreground_background_csr(env, csr_no)                     \
     do{                                                                 \
         target_ulong backup_csr;                                        \
         backup_csr = csr_read_helper(env, CSR_##csr_no);                \
@@ -742,18 +742,17 @@ void set_privilege(CPURISCVState *env, target_ulong newpriv)
         csr_write_helper(env, backup_csr, CSR_##csr_no);                \
     }while(0);
 
-
 void swap_background_background_csrs(CPURISCVState *env){
-    swap_background_background_csr(env, SSTATUS);
-    swap_background_background_csr(env, SIP);
-    swap_background_background_csr(env, SIP);
-    swap_background_background_csr(env, SIE);
-    swap_background_background_csr(env, STVEC);
-    swap_background_background_csr(env, SSCRATCH);
-    swap_background_background_csr(env, SEPC);
-    swap_background_background_csr(env, SCAUSE);
-    swap_background_background_csr(env, STVAL);
-    swap_background_background_csr(env, SATP);
+    swap_foreground_background_csr(env, SSTATUS);
+    swap_foreground_background_csr(env, SIP);
+    swap_foreground_background_csr(env, SIP);
+    swap_foreground_background_csr(env, SIE);
+    swap_foreground_background_csr(env, STVEC);
+    swap_foreground_background_csr(env, SSCRATCH);
+    swap_foreground_background_csr(env, SEPC);
+    swap_foreground_background_csr(env, SCAUSE);
+    swap_foreground_background_csr(env, STVAL);
+    swap_foreground_background_csr(env, SATP);
 }
 
 target_ulong helper_sret(CPURISCVState *env, target_ulong cpu_pc_deb)
-- 
2.7.4

