From c3fd7f5bb6191040506253d3234d971008346c5a Mon Sep 17 00:00:00 2001
From: Zhang Heng <zhangheng@mprc.pku.edu.cn>
Date: Sun, 10 Dec 2017 13:06:42 +0800
Subject: [PATCH 2/2] htif: set mip proper before irq

Signed-off-by: Zhang Heng <zhangheng@mprc.pku.edu.cn>
---
 hw/riscv/htif/htif.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/hw/riscv/htif/htif.c b/hw/riscv/htif/htif.c
index 9be5f02..1861d49 100644
--- a/hw/riscv/htif/htif.c
+++ b/hw/riscv/htif/htif.c
@@ -77,6 +77,7 @@ static void htif_recv(void *opaque, const uint8_t *buf, int size)
     uint64_t resp = 0x100 | *buf;
 
     htifstate->env->mfromhost = (val_written >> 48 << 48) | (resp << 16 >> 16);
+    htifstate->env->mip |= MIP_SEIP;
     qemu_irq_raise(htifstate->irq);
 }
 
@@ -325,6 +326,7 @@ static void htif_mm_write(void *opaque, hwaddr addr,
     } else if (addr == FROMHOST_OFFSET2) {
         htifstate->env->mfromhost |= value << 32;
         if (htifstate->env->mfromhost == 0x0) {
+            htifstate->env->mip &= ~MIP_SEIP;
             qemu_irq_lower(htifstate->irq);
         }
         htifstate->fromhost_inprogress = 0;
-- 
2.7.4

